### PR TITLE
Switch to gengo

### LINKED ISSUE #26: Improve language detection system to recognize C++ headers
https://github.com/github/linguist

Linguist is a tool developed by GitHub for the specific purpose of detecting languages. It's a very mature tool that gets it right the majority of the time by using complex rules.

### PR DESCRIPTION
**Pull request by spenserblack**
See #1152 for previous conversation. With the drastic API changes to gengo it started to feel like more work to modify the existing PR instead of starting over.

# To Do

- [ ] Gengo needs to support the remaining languages (or at least have reasons to *not* support them) (https://github.com/spenserblack/gengo/issues/34)
- [ ] Change pretty much everything that previously matched on `Language` and `tokei::LanguageType` enums.

Resolves #26
Closes #1152

### CODE REVIEW

**Comment by spenserblack** [comment]
@o2sh Before this moves any further we need to determine if we want to move forward with analyzing a git ref (`HEAD`) or a folder (I think preferably we don't support both). I've started listing reasons for each here: https://github.com/o2sh/onefetch/discussions/1303

File: src/info/langs/mod.rs
Comment (o2sh) [inline src/info/langs/mod.rs:1]: *Seems like I cannot add a comment on unchanged lines, so I'm putting this here...* 😓 

You need to update `get_loc_by_language_sorted` method name and description

File: src/info/blob_size.rs
Comment (o2sh) [inline src/info/blob_size.rs:1]: [As discussed in the past](https://github.com/o2sh/onefetch/pull/1152#discussion_r1313829809), I think we can drop this info line and only keep `repo_size`

File: src/info/langs/mod.rs [hunk 1]
```diff
@@ -66,146 +73,13 @@ fn get_statistics(
...
-    let tokei_config = tokei::Config {
-        types: Some(filtered_languages),
-        hidden: Some(include_hidden),
-        ..tokei::Config::default()
-    };
-    let ignored: Vec<&str> = globs_to_exclude.iter().map(AsRef::as_ref).collect();
-    languages.get_statistics(&[&dir], &ignored, &tokei_config);
-    languages
+) -> Result<gengo::Analysis, Box<dyn Error>> {
+    // TODO Determine best way to ignore files (and if that should continue to be handled by onefetch)
```
Comment (o2sh) [inline src/info/langs/mod.rs:77]: It would be nice to have the language type filter and glob filtering for a V1

File: src/info/langs/mod.rs [hunk 2]
```diff
@@ -35,28 +41,29 @@ fn sort_by_loc(map: HashMap<Language, usize>) -> Vec<(Language, usize)> {
     vec
 }
 
-fn get_loc_by_language(languages: &tokei::Languages) -> Option<HashMap<Language, usize>> {
-    let mut loc_by_language = HashMap::new();
+fn get_loc_by_language<Any>(languages: Any) -> Option<HashMap<Language, usize>> {
+    todo!("Determine if this is necessary")
```
Comment (o2sh) [inline src/info/langs/mod.rs:45]: minor: use the `size_by_language logic here`

File: src/info/langs/mod.rs [hunk 3]
```diff
@@ -66,146 +73,13 @@ fn get_statistics(
...
-        types: Some(filtered_languages),
-        hidden: Some(include_hidden),
-        ..tokei::Config::default()
-    };
-    let ignored: Vec<&str> = globs_to_exclude.iter().map(AsRef::as_ref).collect();
-    languages.get_statistics(&[&dir], &ignored, &tokei_config);
-    languages
+) -> Result<gengo::Analysis, Box<dyn Error>> {
+    // TODO Determine best way to ignore files (and if that should continue to be handled by onefetch)
+    let file_source = Git::new(dir, "HEAD")?;
```
Comment (o2sh) [inline src/info/langs/mod.rs:78]: is this going to take into account the pending changes ?

File: src/info/langs/mod.rs (hunk 1 above)
Comment (o2sh) [reply src/info/langs/mod.rs:77]: or is this going to be handled via `.gitattributes`?

Comment (spenserblack) [reply src/info/langs/mod.rs:77]: That depends on the "file source." Right now the `Git` file source is highly opinionated and would use `.gitattributes` for ignoring files, overrides, etc.

******

I plan `Directory` file source being much less opinionated (can't get much more generic than just reading a folder :slightly_smiling_face:). The functionality isn't there yet, but I was considering adding ways to configure excluding/including files. Right now `.gitignore` and `.ignore` files would ignore files via the `ignore` crate's default behavior. I was hoping to discuss the usage sometime, but it would probably either be like this:
```rust
let file_source = Directory::with_config("./", Config { ignored_files: Some(vec![]) });
```
or like this:
```rust
let file_source = Directory::new("./").extend_ignored_files(&[]);
```

File: src/info/langs/mod.rs (hunk 3 above)
Comment (spenserblack) [reply src/info/langs/mod.rs:78]: https://github.com/o2sh/onefetch/discussions/1303#discussioncomment-9057210

File: src/info/langs/mod.rs (hunk 1 above)
Comment (spenserblack) [reply src/info/langs/mod.rs:77]: By default Gengo marks Programming, Markup, and Query as detectable, and Prose and Data as not detectable. `detectable` is basically a boolean that answers "should this factor in to the stats?" Also, documentation, generated files, and vendored files are not detectable by default. Assuming we use the `Git` file source with `.gitattributes`, this is example usage:
```gitattributes
# Markdown is Prose, but we still want it in the stats
*.md gengo-detectable

# JavaScript is Programming, but we want to exclude it
*.js -gengo-detectable

# The contents of dist/ are not generated, so they will be detectable
dist/* -gengo-generated
# OR
# Even though dist/* is generated, it should still be included in the stats
dist/* gengo-detectable
```

******

Again, the `Git` file source is highly opinionated and tries to behave a lot like github-linguist. The usage with the `Directory` file source is still undecided, and right now it doesn't implement any overrides. Or we could write our own file source (https://github.com/o2sh/onefetch/discussions/1303#discussioncomment-9058916) if we want very specific behavior that wouldn't be suitable for the `gengo` crate.

But, however we implement it, IMO the best way would be to inform gengo what files are and aren't detectable, and then make use of either the `Analysis` (detailed results) or `Summary` (simplified results).

******

tl;dr Yes, gengo would be the one handling included/excluded types, and we'd just pass configuration to it somehow.

File: src/info/repo_size.rs
Comment (o2sh) [inline src/info/repo_size.rs:1]: why the renaming?

Comment (spenserblack) [reply src/info/repo_size.rs:1]: Originally because there were two "size" stats (https://github.com/o2sh/onefetch/pull/1305#discussion_r1557311545). I could revert the rename, but I kind of like it for clarifying that we're getting the size of the *repository,* not the size of the code, which can be confusing (https://github.com/o2sh/onefetch/issues/122, https://github.com/o2sh/onefetch/issues/342, https://github.com/o2sh/onefetch/issues/338). Though perhaps this changed with the migration to gix?

File: src/info/langs/mod.rs [hunk 4]
```diff
@@ -19,44 +22,35 @@ pub fn get_loc_by_language_sorted(
     language_types: &[LanguageType],
     include_hidden: bool,
```
Comment (spenserblack) [inline src/info/langs/mod.rs:23]: How do we want to handle including/excluding hidden files? I checked it out and `tokei` basically forwards the configuration to a configuration for `ignore`. `ignore` [has different behavior on Windows](https://github.com/BurntSushi/ripgrep/blob/bb8601b2bafb5e68181cbbb84e6ffa4f7a72bf16/crates/ignore/src/pathutil.rs#L21-L57). As I understand it on Windows it will be hidden if the file's stats say that it's hidden, and on all operating systems it will be hidden if the filename starts with `.`.

`ignore` is much more general-purpose than our usage, as it should work with any directory, not just a repository. So I wonder if we should even handle Windows' special case. Besides adding code complexity, hiding files via Windows file attributes isn't really applicable to git repositories.

File: src/info/repo_size.rs
Comment (o2sh) [reply src/info/repo_size.rs:1]: maybe out of scope for this PR

File: src/info/langs/mod.rs [hunk 5]
```diff
@@ -9,14 +9,14 @@ use strum::IntoEnumIterator;
 
 pub mod language;
 
-pub fn get_main_language(loc_by_language: &[(Language, usize)]) -> Language {
```
Comment (spenserblack) [inline src/info/langs/mod.rs:9]: A bit of refactoring is inevitable to make things make more sense when using `gengo` over `tokei`. Stumbled on this. Don't mean to be harsh, but is a function for the sole purpose of indexing into `[0]` really necessary? Especially now that newer versions of Rust have `.first()` and `.get(0)` to *safely* index into collections.

Comment (o2sh) [reply src/info/langs/mod.rs:9]: Sure, go ahead. 

But for an easier review, we can keep the refactoring/cleanup in a separate PR, so we can focus on the current subject and don't lose track of the important changes.

File: src/info/langs/mod.rs (hunk 4 above)
Comment (spenserblack) [reply src/info/langs/mod.rs:23]: Oh, wait, I just realized the Windows file attributes won't apply anyway if we're reading from the commit instead of the directory.

File: src/info/langs/mod.rs (hunk 5 above)
Comment (spenserblack) [reply src/info/langs/mod.rs:9]: Refactored it to handle a OOB panic that was introduced (see https://github.com/o2sh/onefetch/pull/1305/commits/82da923be01d27b7eed51e0835e7e239582ba512 for more info).

**Comment by spenserblack** [comment]
So I think that covers pretty much everything. All that's left is to add more language support (https://github.com/spenserblack/gengo/issues/34) and make a new release to add back the supported language variants and then we're good AFAIK.

**Comment by spenserblack** [comment]
Alright, after a long time procrastinating, I implemented the last of onefetch's languages (besides combining Bash, Zsh, and Sh into Shell).

A couple things to note:

1. The MSRV has been bumped
2. Right now this brings in a duplicate version of gix. Right now there's a dependency resolution bug and I didn't want to release the gengo lib with a higher version until that's resolved.

**Comment by spenserblack** [comment]
Bumping gix resulting in compiler errors (new usage). Just a moment.

**Comment by spenserblack** [comment]
After thinking about this a bit, I think an issue with tokei and other line-counting tools is that they specialize in counting lines, and, while language ID is a priority, not the *top* priority. Which leads to e.g. tokei having poor out-of-the-box support for Verilog and V (their `*.v` extensions were already taken). So they're not great for properly identifying a language.

So I'd like to add line counts back at some point, possibly by using gengo as a dependency of some crate and doing something vaguely like
```rust
match Language {
    JupyterNotebook => my_super_special_jupyter_notebook_counter()
    _ => generic_counter(Language) // would use community-provided data similar to tokei's languages.json
}
```
